### ISSUE TITLE
crash following memory leak when using bugsnag and Apollo on react native

### ISSUE THREAD

**Issue by otusweb** [bugsnag/bugsnag-js#1606]
### Describe the bug
When passing the Apollo's `Operation.getContext()` result to the Bugsnag's `Event.addMetaData()` method, the memory consumption of the app in iOS shoots through the roof in 5 seconds.
This only reproduce in release build (not debug).

### Steps to reproduce
1. Create a react native app
2. add bugsnag and apollo
3. create an apollo client with a middleware to handle error in which you pass the Operation.getContext() result to bugsnag addMetadata()
4. Run the app in release mode (best way is to click Product > Profile in Xcode)
4. Intercept the traffic to cause a 500 or 404 error when calling the Endpoint

=> Memory usage climbs rapidly until app crashes
expected: app continues to work

### Environment
* Bugsnag version: @bugsnag/core 7.13.2
* Browser framework version (if any):
    * React Native: 0.66.1
    * iOS: 14 & 15
* Device (e.g. iphonex): iphone and iPad 

### Example Repo 

- [ ] Create a minimal repository that can reproduce the issue
- [ ] Link to it here:

### Example code snippet 

```

const logErrorLink = onError(({ graphQLErrors, networkError, operation }) => {

    bugsnagClient.notify(
      new Error(`Error Apollo ${operation.operationName}`), 
      (report) => {
        const _params = {
          data: {
            operationName: operation.operationName,
            extensions: operation.extensions,
            query: operation.query,
            variables: operation.variables,
            context: operation.getContext(),    <=== If i replace this line with context: JSON.stringify(operation.getContext()), the problem goes away
            graphQLErrors,
            networkError,
          } as any,
        };
        report.addMetadata('report', _params);
      }
    );
  });


  const link = ApolloLink.from([
    logErrorLink,
    authMiddleware.concat(persistedQueriesLink.concat(httpLink)),
  ])

  const client = new ApolloClient({
    uri: "https://48p1r2roz4.sse.codesandbox.io",
    cache: new InMemoryCache(),
  })

  return client

```

no specific error message

Here is the memory stack from Instruments
```
   0 libc++abi.dylib operator new(unsigned long)
   1 Choose void* std::__1::__libcpp_operator_new<unsigned long>(unsigned long) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/new:235
   2 Choose std::__1::__libcpp_allocate(unsigned long, unsigned long) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/new:261
   3 Choose std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*> >::allocate(unsigned long) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/memory:870
   4 Choose std::__1::allocator_traits<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*> > >::allocate(std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*> >&, unsigned long) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/__memory/allocator_traits.h:260
   5 Choose std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>, std::__1::__hash_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*> > > > std::__1::__hash_table<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, std::__1::__unordered_map_hasher<folly::dynamic, std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, true>, std::__1::__unordered_map_equal<folly::dynamic, std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, folly::detail::DynamicKeyEqual, folly::detail::DynamicHasher, true>, std::__1::allocator<std::__1::__hash_value_type<folly::dynamic, folly::dynamic> > >::__construct_node<std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>&&, std::__1::tuple<>&&) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/__hash_table:2455
   6 Choose std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>*>, bool> std::__1::__hash_table<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, std::__1::__unordered_map_hasher<folly::dynamic, std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, true>, std::__1::__unordered_map_equal<folly::dynamic, std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, folly::detail::DynamicKeyEqual, folly::detail::DynamicHasher, true>, std::__1::allocator<std::__1::__hash_value_type<folly::dynamic, folly::dynamic> > >::__emplace_unique_impl<std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>&&, std::__1::tuple<>&&) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/__hash_table:2137
   7 Choose std::__1::pair<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>*>, bool> std::__1::__hash_table<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, std::__1::__unordered_map_hasher<folly::dynamic, std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, true>, std::__1::__unordered_map_equal<folly::dynamic, std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, folly::detail::DynamicKeyEqual, folly::detail::DynamicHasher, true>, std::__1::allocator<std::__1::__hash_value_type<folly::dynamic, folly::dynamic> > >::__emplace_unique<std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>&&, std::__1::tuple<>&&) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/__hash_table:1096
   8 Choose std::__1::pair<std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>*> >, bool> std::__1::unordered_map<folly::dynamic, folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::allocator<std::__1::pair<folly::dynamic const, folly::dynamic> > >::emplace<std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>&&, std::__1::tuple<>&&) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/unordered_map:1158
   9 Choose auto std::__1::pair<std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>*> >, bool> folly::f14::detail::F14BasicMap<folly::dynamic, folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::allocator<std::__1::pair<folly::dynamic const, folly::dynamic> > >::emplace<std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>&&, std::__1::tuple<>&&)::'lambda'(auto&, auto&&...)::operator()<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(auto&, auto&&...) const /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/container/detail/F14MapFallback.h:215
  10 Choose _ZN5folly6detail22callWithKeyAndPairArgsINS_7dynamicES2_ZNS_3f146detail11F14BasicMapIS2_S2_NS0_13DynamicHasherENS0_15DynamicKeyEqualENSt3__19allocatorINS8_4pairIKS2_S2_EEEEE7emplaceIJRKNS8_21piecewise_construct_tENS8_5tupleIJRNS8_12basic_stringIcNS8_11char_traitsIcEENS9_IcEEEEEEENSJ_IJEEEEEENSA_INS8_19__hash_map_iteratorINS8_15__hash_iteratorIPNS8_11__hash_nodeINS8_17__hash_value_typeIS2_S2_EEPvEEEEEEbEEDpOT_EUlRT_DpOT0_E_SO_JSP_EJEEEDaOT1_RKT2_ONSJ_IJDpT3_EEEONSJ_IJDpT4_EEE /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/container/detail/Util.h:125
  11 Choose _ZN5folly6detail20callWithExtractedKeyINS_7dynamicES2_NS_3f146detail11F14BasicMapIS2_S2_NS0_13DynamicHasherENS0_15DynamicKeyEqualENSt3__19allocatorINS8_4pairIKS2_S2_EEEEE11UsableAsKeyESD_ZNSE_7emplaceIJRKNS8_21piecewise_construct_tENS8_5tupleIJRNS8_12basic_stringIcNS8_11char_traitsIcEENS9_IcEEEEEEENSK_IJEEEEEENSA_INS8_19__hash_map_iteratorINS8_15__hash_iteratorIPNS8_11__hash_nodeINS8_17__hash_value_typeIS2_S2_EEPvEEEEEEbEEDpOT_EUlRT_DpOT0_E_SQ_JELi0EEEDaRT2_OT3_SH_ONSK_IJT4_EEEONSK_IJDpT5_EEE /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/container/detail/Util.h:155
  12 Choose std::__1::pair<std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>*> >, bool> folly::f14::detail::F14BasicMap<folly::dynamic, folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::allocator<std::__1::pair<folly::dynamic const, folly::dynamic> > >::emplace<std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>, std::__1::tuple<> >(std::__1::piecewise_construct_t const&, std::__1::tuple<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>&&, std::__1::tuple<>&&) /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/container/detail/F14MapFallback.h:203
  13 Choose std::__1::enable_if<folly::detail::EligibleForHeterogeneousInsert<folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>::value, std::__1::pair<std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<folly::dynamic, folly::dynamic>, void*>*> >, bool> >::type folly::f14::detail::F14BasicMap<folly::dynamic, folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::allocator<std::__1::pair<folly::dynamic const, folly::dynamic> > >::try_emplace<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/container/detail/F14MapFallback.h:268
  14 Choose std::__1::enable_if<folly::detail::EligibleForHeterogeneousInsert<folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>::value, folly::dynamic&>::type folly::f14::detail::F14BasicMap<folly::dynamic, folly::dynamic, folly::detail::DynamicHasher, folly::detail::DynamicKeyEqual, std::__1::allocator<std::__1::pair<folly::dynamic const, folly::dynamic> > >::operator[]<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/container/detail/F14MapFallback.h:356
  15 Choose std::__1::enable_if<!(std::is_convertible<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::__wrap_iter<folly::dynamic*> >::value), void>::type folly::dynamic::insert<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::nullptr_t>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::nullptr_t&&) /Users/otusweb/Projects/AppChoose/ios/Pods/Headers/Public/RCT-Folly/folly/dynamic-inl.h:834
  16 Choose facebook::jsi::dynamicFromValue(facebook::jsi::Runtime&, facebook::jsi::Value const&) /Users/otusweb/Projects/AppChoose/node_modules/react-native/ReactCommon/jsi/jsi/JSIDynamic.cpp:187
  17 Choose facebook::react::JSIExecutor::callNativeModules(facebook::jsi::Value const&, bool) /Users/otusweb/Projects/AppChoose/node_modules/react-native/ReactCommon/jsiexecutor/jsireact/JSIExecutor.cpp:420
  18 Choose facebook::react::JSIExecutor::callFunction(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, folly::dynamic const&) /Users/otusweb/Projects/AppChoose/node_modules/react-native/ReactCommon/jsiexecutor/jsireact/JSIExecutor.cpp:270
  19 Choose std::__1::__function::__value_func<void (facebook::react::JSExecutor*)>::operator()(facebook::react::JSExecutor*&&) const /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:1885
  20 Choose std::__1::function<void (facebook::react::JSExecutor*)>::operator()(facebook::react::JSExecutor*) const /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:2560
  21 Choose facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8::operator()() const /Users/otusweb/Projects/AppChoose/node_modules/react-native/ReactCommon/cxxreact/NativeToJsBridge.cpp:310
  22 Choose decltype(std::__1::forward<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8&>(fp)()) std::__1::__invoke<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8&>(facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8&) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/type_traits:3694
  23 Choose void std::__1::__invoke_void_return_wrapper<void, true>::__call<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8&>(facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8&) /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/__functional_base:348
  24 Choose std::__1::__function::__alloc_func<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8, std::__1::allocator<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8>, void ()>::operator()() /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:1558
  25 Choose std::__1::__function::__func<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8, std::__1::allocator<facebook::react::NativeToJsBridge::runOnExecutorQueue(std::__1::function<void (facebook::react::JSExecutor*)>)::$_8>, void ()>::operator()() /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:1732
  26 Choose std::__1::__function::__value_func<void ()>::operator()() const /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:1885
  27 Choose std::__1::function<void ()>::operator()() const /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:2560
  28 Choose facebook::react::tryAndReturnError(std::__1::function<void ()> const&) /Users/otusweb/Projects/AppChoose/node_modules/react-native/React/CxxModule/RCTCxxUtils.mm:74
  29 Choose facebook::react::RCTMessageThread::tryFunc(std::__1::function<void ()> const&) /Users/otusweb/Projects/AppChoose/node_modules/react-native/React/CxxBridge/RCTMessageThread.mm:69
  30 Choose std::__1::__function::__value_func<void ()>::operator()() const /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:1885
  31 Choose std::__1::function<void ()>::operator()() const /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk/usr/include/c++/v1/functional:2560
  32 Choose invocation function for block in facebook::react::RCTMessageThread::runAsync(std::__1::function<void ()>) /Users/otusweb/Projects/AppChoose/node_modules/react-native/React/CxxBridge/RCTMessageThread.mm:45
  33 CoreFoundation __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__
  34 CoreFoundation __CFRunLoopDoBlocks
  35 CoreFoundation __CFRunLoopRun
  36 CoreFoundation CFRunLoopRunSpecific
  37 Choose +[RCTCxxBridge runRunLoop] /Users/otusweb/Projects/AppChoose/node_modules/react-native/React/CxxBridge/RCTCxxBridge.mm:367
  38 Foundation __NSThread__start__
  39 libsystem_pthread.dylib _pthread_start
  40 libsystem_pthread.dylib thread_start
```

The screenshot from instruments highlighting the leak (malloc 112bytes)
<img width="1570" alt="142430451-c2d13a84-55ca-470a-8941-3610f53b5ba4-1" src="https://user-images.githubusercontent.com/1069914/144427248-844ea215-365a-4ef4-8892-a6ff805514dc.png">
:

**Comment by otusweb** [comment]
I finally manage to run this in debug and was able to reproduce in debug too. Once I replace the line with `JSON.stringify(operation.getContext())` it does indeed fix the issue but I get teh following in the logs which is most likely a clue:
```
 ERROR  [bugsnag] Error occurred in onError callback, continuing anyway…
 ERROR  [bugsnag] [TypeError: cyclical structure in JSON object]
```
My gut feeling is that the object returned by getContext is a complex JS object that can't be serialized. 
I wonder if anything can be done to detect this issue instead of trying to serialize it and running into memory issue

**Comment by mattdyoung** [comment]
Hi @otusweb 

Thanks for the info! We agree that this is something we should be able to protect against when we serialize the JSON. We'll take a look at fixing this in a future release.

**Comment by yousif-bugsnag** [comment]
A fix for this (#1673) has now been released in [v7.16.1](https://github.com/bugsnag/bugsnag-js/releases/tag/v7.16.1) 🎉
